File: brain/src/service/persona_selector.py
"""PersonaSelector — LLM-assisted persona selection from installed set.

Chooses which persona(s) an item belongs to by asking the LLM to pick
from the actual installed personas. Never invents persona names.

Resolution order:
1. Explicit valid persona_hint → use it
2. Constrained LLM selection → choose from installed list only
3. Validate → drop anything not in registry
4. Fallback → default persona
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any

import structlog

from ..prompts import PROMPT_PERSONA_CLASSIFY_SYSTEM as _SYSTEM_PROMPT
from .persona_registry import PersonaInfo, PersonaRegistry

log = structlog.get_logger(__name__)


@dataclass
class SelectionResult:
    """Result of persona selection."""

    primary: str | None = None
    secondary: list[str] = field(default_factory=list)
    confidence: float = 0.0
    reason: str = ""
    has_event: bool = False
    event_hint: str = ""
    attribution_corrections: list[dict] = field(default_factory=list)
    # LLM corrections to deterministic attributions, by stable ID:
    # [{"id": 1, "corrected_bucket": "self_explicit", "reason": "..."}]


class PersonaSelector:
    """Selects persona(s) for an item using constrained LLM selection.

    Parameters
    ----------
    registry:
        PersonaRegistry for validation.
    llm:
        LLM router for classification calls.
    default_persona:
        Fallback persona when LLM is unavailable or low-confidence.
    """

    def __init__(
        self,
        registry: PersonaRegistry,
        llm: Any = None,
        default_persona: str = "general",
    ) -> None:
        self._registry = registry
        self._llm = llm
        self._default = default_persona

    async def select(
        self,
        item: dict,
        persona_hint: str | None = None,
    ) -> SelectionResult | None:
        """Select persona(s) for an item.

        1. If persona_hint is valid, use it directly.
        2. Otherwise, ask LLM to choose from installed personas.
        3. Validate result against registry.
        4. Return None if no confident selection — caller uses deterministic fallback.
        """
        # 1. Explicit valid hint
        if persona_hint:
            norm = self._registry.normalize(persona_hint)
            if self._registry.exists(norm):
                return SelectionResult(
                    primary=norm,
                    confidence=1.0,
                    reason="explicit persona hint",
                )

        # 2. LLM selection
        if self._llm:
            try:
                result = await self._llm_select(item)
                if result and result.primary:
                    return result
            except Exception as exc:
                log.warning("persona_selector.llm_failed", error=str(exc))

        # No confident answer — return None so caller can use deterministic fallback
        return None

    async def _llm_select(self, item: dict) -> SelectionResult | None:
        """Ask the LLM to choose from installed personas."""
        personas = self._registry.all_names()
        if not personas:
            return None

        # Build persona descriptions for the prompt — include description
        # from the registry so the LLM knows what each persona is for.
        persona_list = []
        for name in personas:
            tier = self._registry.tier(name) or "default"
            desc = self._registry.description(name)
            entry: dict = {"name": name, "tier": tier}
            if desc:
                entry["description"] = desc
            persona_list.append(entry)

        # Build item context (scrub to essentials)
        item_context = {
            "item_type": item.get("type", ""),
            "source": item.get("source", ""),
            "sender": item.get("sender", ""),
            "summary": (item.get("summary", "") or "")[:200],
            "body_preview": (item.get("body", "") or "")[:300],
        }

        # Include mentioned contacts for relationship-aware routing.
        mentioned = item.get("mentioned_contacts")
        if mentioned:
            item_context["mentioned_contacts"] = mentioned

        # Include attribution candidates for LLM correction.
        attribution_candidates = item.get("attribution_candidates")
        if attribution_candidates:
            item_context["attribution_candidates"] = attribution_candidates

        import datetime as _dt
        today = _dt.datetime.now(_dt.timezone.utc).strftime("%Y-%m-%d")

        prompt = json.dumps({
            "today": today,
            "available_personas": persona_list,
            **item_context,
        }, indent=2)

        messages = [
            {"role": "system", "content": _SYSTEM_PROMPT},
            {"role": "user", "content": prompt},
        ]

        resp = await self._llm.route(
            task_type="classification",
            prompt=prompt,
            messages=messages,
        )

        content = resp.get("content", "")
        return self._parse_response(content)

    def _parse_response(self, content: str) -> SelectionResult | None:
        """Parse LLM JSON response and validate against registry."""
        try:
            # Extract JSON from possible markdown code fences
            text = content.strip()
            if text.startswith("```"):
                text = text.split("\n", 1)[-1].rsplit("```", 1)[0].strip()

            data = json.loads(text)

            primary = data.get("primary", "")
            secondary = data.get("secondary", [])
            confidence = float(data.get("confidence", 0.0))
            reason = data.get("reason", "")

            # Validate primary
            if primary and not self._registry.exists(primary):
                log.warning(
                    "persona_selector.invalid_primary",
                    primary=primary,
                    available=self._registry.all_names(),
                )
                return None

            # Filter secondary to valid personas only
            valid_secondary = [
                s for s in secondary
                if s != primary and self._registry.exists(s)
            ]

            return SelectionResult(
                primary=primary or None,
                secondary=valid_secondary,
                confidence=confidence,
                reason=reason,
                has_event=bool(data.get("has_event", False)),
                event_hint=data.get("event_hint", ""),
                attribution_corrections=data.get("attribution_corrections", []),
            )
        except (json.JSONDecodeError, ValueError, KeyError) as exc:
            log.warning("persona_selector.parse_failed", error=str(exc))
            return None

File: apps/home-node-lite/core-server/src/brain/persona_selector.ts
/**
 * Task 5.44 — persona selector.
 *
 * Decides which persona(s) a captured item belongs to. Never invents
 * persona names — the LLM is given the actual installed set (from
 * PersonaRegistry) and told to pick among those.
 *
 * **Resolution order**:
 *   1. Explicit valid `personaHint` → use directly (confidence=1.0).
 *   2. LLM selection — constrained to installed personas.
 *   3. Validate — drop anything not in registry.
 *   4. Return `null` if no confident answer — caller uses its
 *      deterministic fallback (typically "default persona").
 *
 * **Never throws**: every failure path produces a structured result
 * or `null`. LLM errors are swallowed + logged via `onEvent`.
 *
 * **Pluggable LLM**: the selector takes an `llmSelectFn` — the
 * caller wires this to ModelRouter with the constrained-JSON schema
 * from `prompts/persona_classify`. Tests pass scripted stubs.
 *
 * Source: docs/HOME_NODE_LITE_TASKS.md Phase 5e task 5.44.
 */

import type { PersonaRegistry } from './persona_registry';

export interface SelectableItem {
  type?: string;
  source?: string;
  sender?: string;
  summary?: string;
  body?: string;
  mentionedContacts?: string[];
  attributionCandidates?: unknown;
}

export interface SelectionResult {
  primary: string | null;
  secondary: string[];
  /** 0..1 — how confident the selector is. 1.0 for explicit hints. */
  confidence: number;
  reason: string;
  hasEvent: boolean;
  eventHint: string;
}

/**
 * Raw LLM response shape. Caller validates each field; `null` or
 * garbage collapses to "no confident answer".
 */
export interface PersonaLlmResponse {
  primary?: string;
  secondary?: string[];
  confidence?: number;
  reason?: string;
  has_event?: boolean;
  event_hint?: string;
}

/**
 * LLM call. Takes an item-context payload + the list of valid
 * persona names + tiers. Returns a structured response or throws.
 */
export type PersonaLlmSelectFn = (ctx: {
  availablePersonas: Array<{ name: string; tier: string; description?: string }>;
  item: SelectableItem;
}) => Promise<PersonaLlmResponse>;

export type PersonaSelectorEvent =
  | { kind: 'explicit_hint'; persona: string }
  | { kind: 'llm_selected'; primary: string; secondary: string[]; confidence: number }
  | { kind: 'llm_failed'; error: string }
  | { kind: 'invalid_primary'; primary: string; available: string[] }
  | { kind: 'no_confident_answer' };

export interface PersonaSelectorOptions {
  registry: PersonaRegistry;
  /** Optional LLM. When absent, only the explicit-hint path runs. */
  llmSelectFn?: PersonaLlmSelectFn;
  /** Diagnostic hook. */
  onEvent?: (event: PersonaSelectorEvent) => void;
}

export class PersonaSelector {
  private readonly registry: PersonaRegistry;
  private readonly llmSelectFn?: PersonaLlmSelectFn;
  private readonly onEvent?: (event: PersonaSelectorEvent) => void;

  constructor(opts: PersonaSelectorOptions) {
    if (!opts.registry) {
      throw new TypeError('PersonaSelector: registry is required');
    }
    this.registry = opts.registry;
    if (opts.llmSelectFn) this.llmSelectFn = opts.llmSelectFn;
    this.onEvent = opts.onEvent;
  }

  /**
   * Select a persona for `item`. Returns `null` when no confident
   * answer is available — the caller then falls back to the default
   * persona (usually "general").
   */
  async select(
    item: SelectableItem,
    personaHint?: string | null,
  ): Promise<SelectionResult | null> {
    // 1. Explicit valid hint wins.
    if (typeof personaHint === 'string' && personaHint.length > 0) {
      const normalised = this.registry.normalize(personaHint);
      if (this.registry.exists(normalised)) {
        this.emit({ kind: 'explicit_hint', persona: normalised });
        return {
          primary: normalised,
          secondary: [],
          confidence: 1.0,
          reason: 'explicit persona hint',
          hasEvent: false,
          eventHint: '',
        };
      }
    }

    // 2. LLM selection.
    if (!this.llmSelectFn) {
      this.emit({ kind: 'no_confident_answer' });
      return null;
    }

    const names = this.registry.allNames();
    if (names.length === 0) {
      this.emit({ kind: 'no_confident_answer' });
      return null;
    }

    const availablePersonas = names.map((name) => {
      const description = this.registry.description(name);
      const entry: { name: string; tier: string; description?: string } = {
        name,
        tier: this.registry.tier(name) ?? 'default',
      };
      if (description.length > 0) entry.description = description;
      return entry;
    });

    let raw: PersonaLlmResponse;
    try {
      raw = await this.llmSelectFn({ availablePersonas, item: trimItem(item) });
    } catch (err) {
      const msg = err instanceof Error ? err.message : String(err);
      this.emit({ kind: 'llm_failed', error: msg });
      return null;
    }

    return this.validateResponse(raw);
  }

  private validateResponse(raw: PersonaLlmResponse): SelectionResult | null {
    if (!raw || typeof raw !== 'object') {
      this.emit({ kind: 'no_confident_answer' });
      return null;
    }
    const primaryRaw = typeof raw.primary === 'string' ? raw.primary.trim() : '';
    if (primaryRaw === '') {
      this.emit({ kind: 'no_confident_answer' });
      return null;
    }
    const primary = this.registry.normalize(primaryRaw);
    if (!this.registry.exists(primary)) {
      this.emit({
        kind: 'invalid_primary',
        primary: primaryRaw,
        available: this.registry.allNames(),
      });
      return null;
    }

    const secondaryIn = Array.isArray(raw.secondary) ? raw.secondary : [];
    const secondary: string[] = [];
    const seen = new Set<string>([primary]);
    for (const s of secondaryIn) {
      if (typeof s !== 'string') continue;
      const norm = this.registry.normalize(s.trim());
      if (norm === '' || seen.has(norm)) continue;
      if (!this.registry.exists(norm)) continue;
      secondary.push(norm);
      seen.add(norm);
    }

    const confidence =
      typeof raw.confidence === 'number' && Number.isFinite(raw.confidence)
        ? clamp01(raw.confidence)
        : 0;
    const reason = typeof raw.reason === 'string' ? raw.reason : '';
    const hasEvent = raw.has_event === true;
    const eventHint = typeof raw.event_hint === 'string' ? raw.event_hint : '';

    this.emit({
      kind: 'llm_selected',
      primary,
      secondary,
      confidence,
    });
    return { primary, secondary, confidence, reason, hasEvent, eventHint };
  }

  private emit(event: PersonaSelectorEvent): void {
    this.onEvent?.(event);
  }
}

// ── Internals ──────────────────────────────────────────────────────────

/**
 * Trim `item` to its LLM-relevant essentials — keeps the prompt
 * short + prevents accidentally spilling long body content through
 * the selector. Body + summary are truncated to Python-reference
 * sizes (200/300 chars).
 */
function trimItem(item: SelectableItem): SelectableItem {
  const out: SelectableItem = {};
  if (item.type) out.type = item.type;
  if (item.source) out.source = item.source;
  if (item.sender) out.sender = item.sender;
  if (typeof item.summary === 'string' && item.summary.length > 0) {
    out.summary = item.summary.slice(0, 200);
  }
  if (typeof item.body === 'string' && item.body.length > 0) {
    out.body = item.body.slice(0, 300);
  }
  if (Array.isArray(item.mentionedContacts) && item.mentionedContacts.length > 0) {
    out.mentionedContacts = [...item.mentionedContacts];
  }
  if (item.attributionCandidates !== undefined) {
    out.attributionCandidates = item.attributionCandidates;
  }
  return out;
}

function clamp01(n: number): number {
  if (n < 0) return 0;
  if (n > 1) return 1;
  return n;
}
